### PR TITLE
Add WorkProcessor#flatten method

### PR DESCRIPTION
**Pull request by sopel39**
This method enables flattening of WorkProcessors
    created directly by WorkProcessor#create
    It is also less complex then previous flatTransform
    implementation

### CODE REVIEW

File: presto-main/src/main/java/com/facebook/presto/operator/WorkProcessorUtils.java [hunk 1]
```diff
@@ -162,42 +162,31 @@ else if (processor.isBlocked()) {
...
+    }
 
-                    if (processor.process()) {
-                        if (!processor.isFinished()) {
-                            return ProcessorState.ofResult(processor.getResult(), false);
-                        }
+    static <T> WorkProcessor<T> flatten(WorkProcessor<WorkProcessor<T>> processor)
+    {
+        requireNonNull(processor, "processor is null");
+        return processor.transform(nestedProcessorOptional -> {
```
Comment (findepi) [inline presto-main/src/main/java/com/facebook/presto/operator/WorkProcessorUtils.java]: Document Optional's meaning at `com.facebook.presto.operator.WorkProcessor.Transformation#process`

File: presto-main/src/main/java/com/facebook/presto/operator/WorkProcessorUtils.java [hunk 2]
```diff
@@ -162,42 +162,31 @@ else if (processor.isBlocked()) {
...
-                    else if (processor.isBlocked()) {
-                        return ProcessorState.blocked(processor.getBlockedFuture());
-                    }
-                    else {
-                        return ProcessorState.yield();
-                    }
+            WorkProcessor<T> nestedProcessor = nestedProcessorOptional.get();
+            if (nestedProcessor.process()) {
+                if (nestedProcessor.isFinished()) {
+                    return ProcessorState.needsMoreData();
```
Comment (findepi) [inline presto-main/src/main/java/com/facebook/presto/operator/WorkProcessorUtils.java]: Why? The nested processor didn't need more data. We 
Add javadocs in ProcessorState and/or Transformation and/or `com.facebook.presto.operator.WorkProcessor#transform`

File: presto-main/src/main/java/com/facebook/presto/operator/WorkProcessorUtils.java [hunk 3]
```diff
@@ -162,42 +162,31 @@ else if (processor.isBlocked()) {
...
                 }
+
+                return ProcessorState.ofResult(nestedProcessor.getResult(), false);
```
Comment (findepi) [inline presto-main/src/main/java/com/facebook/presto/operator/WorkProcessorUtils.java]: will `nestedProcessor` be fed back as `nestedProcessorOptional` next time we're invoked?

File: presto-main/src/main/java/com/facebook/presto/operator/WorkProcessorUtils.java [hunk 4]
```diff
@@ -72,7 +72,8 @@ protected T computeNext()
 
                     return Optional.of(processor.getResult());
                 }
-                else if (processor.isBlocked()) {
```
Comment (findepi) [inline presto-main/src/main/java/com/facebook/presto/operator/WorkProcessorUtils.java]: cmt msg
> Remove redundant else keyword

remove "keyword"

File: presto-main/src/main/java/com/facebook/presto/operator/WorkProcessorUtils.java [hunk 5]
```diff
@@ -162,42 +163,31 @@ else if (processor.isBlocked()) {
...
-                    }
-                    else if (processor.isBlocked()) {
-                        return ProcessorState.blocked(processor.getBlockedFuture());
-                    }
-                    else {
-                        return ProcessorState.yield();
-                    }
+            WorkProcessor<T> nestedProcessor = nestedProcessorOptional.get();
+            if (nestedProcessor.process()) {
+                if (nestedProcessor.isFinished()) {
```
[the remark this answers is not in the document]
Comment (sopel39) [reply presto-main/src/main/java/com/facebook/presto/operator/WorkProcessorUtils.java]: You can have nested work processors, e.g:
```
{{1,2,3}, {4,5,6}}
```
In such case, `processor={{1,2,3}, {4,5,6}}`, but nested work processors would be:
```
nested_work_processor1={1,2,3}
nested_work_processor2={4,5,6}
```
after `nested_work_processor1` is finished, there is still `nested_work_processor2` to be processed. Thus it is requested via `needsMoreData()`.

Overall, what `flatten` does it it converts `WorkProcessor<WorkProcessor<T>>` into `WorkProcessor<T>`.

[the remark this answers is not in the document]
Comment (sopel39) [reply presto-main/src/main/java/com/facebook/presto/operator/WorkProcessorUtils.java]: It is similar to: https://google.github.io/guava/releases/21.0/api/docs/com/google/common/collect/Iterables.html#concat-java.lang.Iterable-

File: presto-main/src/main/java/com/facebook/presto/operator/WorkProcessor.java [hunk 6]
```diff
@@ -127,6 +127,33 @@
...
+        /**
+         * Processes input elements and returns current transformation state.
+         *
+         * @param elementOptional an element to be transformed. Will be empty
+         * when there are no more elements. In such case transformation should
+         * finish processing and flush any remaining data.
+         * @return a current transformation state.
+         * <ul>
+         * <li>if state is {@link ProcessorState#finished()} then transformation has finished
+         * and {@link Transformation#process(java.util.Optional)} won't be called again;</li>
```
Comment (findepi) [inline presto-main/src/main/java/com/facebook/presto/operator/WorkProcessor.java]: nit: `s/{@link Transformation#process(java.util.Optional)}/this method/` or `the process method` (here & below)

nit: no need for `;` at the end of list items

File: presto-main/src/main/java/com/facebook/presto/operator/WorkProcessor.java [hunk 7]
```diff
@@ -127,6 +127,33 @@
...
+         * <li>if state is {@link ProcessorState#needsMoreData} then transformation requires
```
Comment (findepi) [inline presto-main/src/main/java/com/facebook/presto/operator/WorkProcessor.java]: missing `()` (currently this links to a field, not the method)

File: presto-main/src/main/java/com/facebook/presto/operator/WorkProcessor.java [hunk 8]
```diff
@@ -127,6 +127,33 @@
...
+         * {@link Transformation#process(java.util.Optional)} will be called again with
+         * the same input element;</li>
+         * <li>if state is {@link ProcessorState#blocked} then transformation is blocked.
+         * {@link Transformation#process(java.util.Optional)} will be called again with
+         * the same input element after blocked future is done;</li>
+         * <li>if state is {@link ProcessorState#ofResult(Object, boolean)} then transformation
+         * has produces a result. {@link Transformation#process(java.util.Optional)} will be called
+         * again with a new element (or with {@link Optional#empty()} if there are no more elements)
+         * if <code>needsMoreData</code> {@link ProcessorState#ofResult(Object, boolean)} parameter
+         * is <code>true</code>.</li>
```
Comment (findepi) [inline presto-main/src/main/java/com/facebook/presto/operator/WorkProcessor.java]: When you read this, you cannot be sure that you parsed this sentence the way it was intended to.. Split this into 2 list items (or sub-items), depending on the `needsMoreData` value.

Also, `com.facebook.presto.operator.WorkProcessor.ProcessorState#ofResult(T, boolean)` doesn't actually represent a processing state. It represents a transition (outputting value) to a state. `ProcessorState` should be renamed to `ProcessorResult` and `ProcessorState#Type` should be renamed to `ProcessorState`.
(might be as a follow-up)

File: presto-main/src/main/java/com/facebook/presto/operator/WorkProcessor.java [hunk 9]
```diff
@@ -127,6 +127,33 @@
...
+         * <li>if state is {@link ProcessorState#finished()} then transformation has finished
+         * and {@link Transformation#process(java.util.Optional)} won't be called again;</li>
+         * <li>if state is {@link ProcessorState#needsMoreData} then transformation requires
+         * more data in order to continue. {@link Transformation#process(java.util.Optional)}
+         * will be called with a new input element or with {@link Optional#empty()} if there
+         * are no more elements;</li>
+         * <li>if state is {@link ProcessorState#yield()} then transformation has yielded.
+         * {@link Transformation#process(java.util.Optional)} will be called again with
+         * the same input element;</li>
+         * <li>if state is {@link ProcessorState#blocked} then transformation is blocked.
```
Comment (findepi) [inline presto-main/src/main/java/com/facebook/presto/operator/WorkProcessor.java]: missing `()` (currently this links to a field, not the method)

**Pull request merged**